src/Microsoft.Extensions.Configuration.AzureAppConfiguration/AzureAppConfigurationProvider.cs

```diff
@@ -420,7 +420,7 @@ public async Task<bool> TryRefreshAsync(CancellationToken cancellationToken)
 
                 return false;
             }
-            catch (AggregateException e) when (e?.InnerExceptions?.All(e => e is RequestFailedException) ?? false)
+            catch (AggregateException e) when (e?.InnerExceptions?.All(e => e is RequestFailedException || e is OperationCanceledException) ?? false)
```

Comment:
We are handling AggregateException where all inner exceptions are OperationCanceledException. This might not be the intended behavior. I think we should handle AggregateException only when at least one inner exception is RequestFailedException. As for other inner exceptions, I'm conflicted about handling OperationCanceledException explicitly vs handling AggregateException with any inner exception as long as there is at least one RequestFailedException. 
We got a customer issue where they saw AggregateException with 2 RequestFailedException and 1 OperationCanceledException. I dont know if there are any other exceptions that SDK might throw inside AggregateException, but for future-proofing the code, I feel like we should handle every AggregateException which has at least one RequestFailedException. 

@jimmyca15 thoughts?

Reply:
My thoughts --

We handle RequestFailedException, OperationCanceledException, KeyVaultReferenceException.

AggregateException is a holder of exceptions that we have to account for. Usually the way I see it handled is to delegate to known inner exception type. So what that would mean here is that if AggregateException contains any of the exceptions we know (RequestFailedException, OperationCanceledException, KeyVaultReferenceException), which we are already swallowing, then we should swallow it as well.

Reply:
Also, I would intentionally order AggregateException as the last caught exception type.

Reply:
Something like

```
catch (AggregateException ae)
{
    if (ae.Any(e is OperationCanceledException))
    {
        _logger.LogWarning(LogHelper.BuildRefreshCanceledErrorMessage());
    }
    else if (ae.Any(e is RequestFailedException))
    {
        if (IsAuthenticationError(e))
        {
            _logger.LogWarning(LogHelper.BuildRefreshFailedDueToAuthenticationErrorMessage(e.Message));
        }
        else
        {
            _logger.LogWarning(LogHelper.BuildRefreshFailedErrorMessage(e.Message));
        }
    }
    else if (ae.Any(e is KeyVaultReferenceException))
    {
        _logger.LogWarning(LogHelper.BuildRefreshFailedDueToKeyVaultErrorMessage(e.Message));
    }
    else
    {
        throw;
    }
}
```

Reply:
We dont need to worry about KeyVaultReferenceException here because that's our internal exception, so it would never be wrapped inside AggregateException. The logic looks good otherwise.

Reply:
In general that's how I would expect aggregate exception to look regardless of possibilities. It is merely possible holder and is delegated to exceptions we already handle.

Reply:
When we catch an AggregateException, we usually just care about the RequestFailedException inside it to determine if its an auth error or if its a failover-able error. We dont really care about operation cancelled if its accompanied by another RequestFailedException (right?). So keeping the complete AggregateException handling pattern may not be necessary in our case. We can trim it down to something more concise:

```
catch (RequestFailedException rfe)
{
}
catch (KeyVaultReferenceException kvre)
{
}
catch (OperationCanceledException oce)
{
}
catch (AggregateException ae) when (ae?.InnerExceptions?.Any(e => e is RequestFailedException || e is OperationCanceledException) ?? false)
{
}
```

And the helper methods `IsAuthenticationError()` and `IsFailOverable()` can be updated to find the RequestFailedException inside AggregateException and log appropriately.

Reply:
Inside the catch you put, are you expecting to have 

```

            if (ae.InnerExceptions.Any(e => e is e is OperationCanceledException))
            {
                _logger.LogWarning(LogHelper.BuildRefreshCanceledErrorMessage());
            }
```

?

If so, I'd avoid using the `when` clause to avoid repetition.

Reply:
I was not expecting the AggregateException catch block to have any changes. When we catch an AggregateException here, its either an authentication error or generic refresh failed error.

Reply:
Well that doesn't seem to add up for the case where AggregateException only has an OperationCanceledException in it. The handling wouldn't be the same as the catch block for OperationCanceledException.

Reply:
Ah I see. Okay, so we can remove the when clause and handle inner exceptions inside the catch block. 
We should still make sure that RequestFailedException is given preference over OperationCancelledException when both are present as inner exceptions.

Reply:
I'm fine with it.